docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+    2 | 3 | 5 | 7 | 11 => println!("This is a prime"),
+    // Match an inclusive range
+    13..=19 => println!("A teen"),
+    // Handle the rest of cases
+    _ => println!("Ain't special"),
+}
+```
+
+Another nice property of match is that it returns a value, which tied to the fact that it must cover all cases means you get a readable idiom for ad-hoc matching and mapping:
+
+```
+let x = match number {
+    0     => "Zero",
+    1..=9 => "Some",
+    _     => "Many",
+}
+```
+
+Obviously where an if-statement is simply there to do optional work, and not cover every case, it may well be more suitable to just use that.
+
+# Error Handling
+
+See also the [Error Handling](https://doc.rust-lang.org/book/ch09-00-error-handling.html) chapter in the Rust Book.
+
+The ICU4X library should be designed so as to never fail at runtime. Now obviously that's something you'd expect all library writers to say, but in Rust you can control the places where code can fail explicitly, so it's much easier to write self-documenting code where the assumptions around failure are obvious.
+
+Most core rust APIs (traits) have two ways to access data, a version that can "panic" (equivalent to a SEGFAULT in C/C++), and one which returns a [Result](https://doc.rust-lang.org/std/result/index.html) or an [Option](https://doc.rust-lang.org/std/option/enum.Option.html). Rust is a language which likes to avoid "unnecessary" overhead, and in some cases it is perfectly correct to use an API which can "panic" because you have already checked the arguments carefully (and performance will be better).
+
+Note that in cases where the Rust compiler can statically determine that a check is sufficient to avoid panic, it will remove the internal check and panic related code, leaving just a provably safe data access.
+
+**Open Question**: Should we have a standard `IcuResult<T>` type (e.g. `Result<T, &'static str>`, or something which could later provide internationalized error messages; e.g. an enum per error).
+
+## Panicking APIs
+
+The most common example of an API which can panic is access to slices and elements of a slice. This includes accessing `array`, `str`, `Vec`, `HashMap` etc. via the `[`,`]` (square bracket) operator, implemented by the [Index](https://doc.rust-lang.org/std/ops/trait.Index.html) trait.
+
+Thus statements like:
+
+```
+let x = self.data[n];
+```
+
+are always prone to "panic" if the index/key is incorrect. (⚠️ and this includes accessing maps ⚠️).
+
+This is because these APIs return a value (or value reference) which cannot be "null", so there is no way for them to signal failure via the return type.
+
+## Non-Panicking APIs
+
+The alternative to using direct data accessors which can panic is to use a method which can return **Option** or **Result**. In the case of collections and strings, where a simple data item is being requested, this is most often provided by functions such as "get" (or "get_mut" for mutable references) which return an Option.
```

Comment:
Something related: does the existence of `Option` return for `get` make `contains()` unnecessary?

Reply:
Why? `contains` is cheaper than `get` because it doesn't care about making the data ready for public consumption in the `exists` scenario, and doesn't care about lifetimes.